### DISCUSSION TITLE
upgrade is not working for custom tap

### DISCUSSION THREAD

**Discussion by devton-here** [Homebrew/discussions#3303] [Getting started] [answered]
1. Initial installation works fine (I use an specific and old version to test upgrade later) `brew install devtonhere/bottle-test/olp@10.5.125`
2. When I try to use `upgrade` it doesn't work: `brew upgrade devtonhere/bottle-test/olp`:
> Error: devtonhere/bottle-test/olp@11.1.0 not installed
3. I can manually specify the version, then it's working (except relinking): ` brew install devtonhere/bottle-test/olp@11.1.0`
After `brew link --overwrite olp@11.1.0` it's working as I would expect.

I would expect that `brew upgrade my-formula` will automatically use latest version (via the symlink in the Alias folder of my tap) to install it and also automatically update the **brew link**.

How can I get it working?

**Comment by SMillerDev** [comment, marked as answer]
So here is what brew is actually doing:

`brew install devtonhere/bottle-test/olp@10.5.125`: install `olpAT105125.rb`
`brew upgrade devtonhere/bottle-test/olp`: resolve the symlink to `olpAT1110.rb` and find that that isn't installed, error out

As far as brew is concerned that might as well be `brew install ruby` and `brew install nodejs` since they're not actually the same software. This is why you'll find that the rules for brew versioned formula specify that it can only differ in minor versions since then you can still have `php@7.4` (7.4.1) -> `php@7.4` (7.4.2) updates.

What you are trying to do with versions is simply not a supported usage of brew. Supported use cases would be:

- have an `olp` formula that you keep updating
- have an `olp` formula with an `olp@11` alias and `olp@10` as a separate formula that users can only switch between using `install`, `link` and `uninstall`

**Reply by devton-here** [reply]
Okay, I see. 

So first option is to have just a single formula file, right? And just updating the version within the `url` and the `version` variable?
The filename would be always `olp.rb` and the class would also be always `Olp`?

That would mean a user could not install an old version unless he is checking out an old revision of the tap, right?

**Reply by SMillerDev** [reply]
Yes, that is the general approach homebrew takes.
